packages/core/src/browser/frontend-application.ts

```diff
@@ -224,6 +255,45 @@ export class FrontendApplication {
         }
     }
 
+    /**
+     * `beforeunload` listener implementation
+     */
+    protected getBeforeUnloadHandler() {
+        return (event: BeforeUnloadEvent) => {
+            const confirmExit = this.corePreferences['application.confirmExit'];
+            let preventExit = false;
+
+            const onWillStopEvent: OnWillStopEvent = {
+                app: this,
+                preventStop() {
+                    preventExit = true;
+                },
+                onStay(callback) {
+                    // `beforeunload` event runs on 1 tick, every else is paused
+                    // This callback should be called once the dialog is closed.
+                    setTimeout(callback);
+                }
+            };
+            for (const contribution of this.contributions.getContributions()) {
+                if (contribution.onWillStop) {
+                    contribution.onWillStop(onWillStopEvent);
+                }
+            }
+            if (confirmExit === 'never') {
+                return;
+            }
+            if (confirmExit === 'always') {
+                preventExit = true;
+            }
+            if (preventExit) {
+                const e = (event || window.event);
+                e.returnValue = 'Application prevented the exit event.';
```

Comment:
`returnValue` is old API, modern browsers don't allow custom values because of security issues, it can be always an empty string to satisfy chrome